extensions/iceberg/s3/src/main/java/io/deephaven/iceberg/util/IcebergToolsS3.java

```diff
@@ -133,10 +139,11 @@ public static IcebergCatalogAdapter createAdapter(
             @NotNull final S3Instructions instructions) {
         final Map<String, String> newProperties = new HashMap<>(properties);
         final Runnable cleanup = DeephavenAwsClientFactory.addToProperties(instructions, newProperties);
-        try {
-            return IcebergTools.createAdapter(name, newProperties, hadoopConfig);
-        } finally {
-            cleanup.run();
-        }
+        final IcebergCatalogAdapter adapter = IcebergTools.createAdapter(name, newProperties, hadoopConfig);
+        // When the Catalog becomes unreachable, we can invoke the DeephavenAwsClientFactory cleanup.
+        // Note: it would be incorrect to register the cleanup against the adapter since the Catalog can outlive the
+        // adapter (and the DeephavenAwsClientFactory properties are needed by the Catalog).
+        CLEANER.register(adapter.catalog(), cleanup);
+        return adapter;
```

Comment:
When I had to do something similar for the `S3Request` objects, Ryan suggested using `CleanupReferenceProcessor`. You can check that too, if that has any advantages.

Reply:
The "register" method on Cleaner is very nice. I tried to add a similarly helpful register method to CleanupReferenceProcessor (essentially, creating a reference behind the scenes that ties an object and a cleanup action), and it _almost_ worked... for some reason though, the caller needs to explicitly retain the returned reference, whereas the same limitation does not apply to Cleaner. It could be I was missing some subtle aspect of the Reference stuff - will have convo w/ Ryan.

Reply:
Cool.

Reply:
I've got a separate PR that adds similar functionality to CleanupReferenceProcessor; https://github.com/deephaven/deephaven-core/pull/6213